### PR TITLE
Apt: Add apt_preference resource from apt cookbooks

### PR DESCRIPTION
**Pull request by tas50**
### Description

Pull in the apt_preference provider from the apt cookbook. This fully deprecates the apt cookbook.

### Issues Resolved
N/A

### Check List

- [ ] New functionality includes tests
- [ ] All tests pass
- [ ] RELEASE\_NOTES.md, has been updated if required (not required for bugfixes, required for API changes)
- [ ] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

Specs are a placeholder on the provider side

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

File: lib/chef/provider/apt_preference.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+class Chef
+  class Provider
+    class AptPreference < Chef::Provider
+      use_inline_resources
+
+      extend Chef::Mixin::Which
+
+      provides :apt_preference do
+        which("apt-get")
+      end
```
Comment (lamont-granquist) [inline lib/chef/provider/apt_preference.rb]: this is probably going to do the same thing where it binds super tightly and overrides the cookbook.

Comment (tas50) [inline lib/chef/provider/apt_preference.rb]: That's fine. The cookbook is going to go away

**Comment by tas50** [comment]
@chef/client-core This is ready for review

**Comment by thommay** [comment]
failing tests…

File: lib/chef/provider/apt_preference.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      action :add do
+        preference = build_pref(
+          new_resource.glob || new_resource.package_name,
+          new_resource.pin,
+          new_resource.pin_priority
+        )
+
+        declare_resource(:directory, APT_PREFERENCE_DIR) do
+          mode "0755"
+          recursive true
```
Comment (coderanger) [inline lib/chef/provider/apt_preference.rb]: This makes me very uncomfortable.

**Comment by tas50** [comment]
@chef/client-core This is ready for review

**Comment by thommay** [comment]
appveyor is angry

File: lib/chef/resource/apt_preference.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+# limitations under the License.
+#
+
+require "chef/resource"
+
+class Chef
+  class Resource
+    class AptPreference < Chef::Resource
+      resource_name :apt_preference
+      provides :apt_preference
```
Comment (coderanger) [inline lib/chef/resource/apt_preference.rb]: Technically redundant but doesn't hurt anything.

**Pull request merged**